### PR TITLE
Implements "Social network" as a new referrer type

### LINKED ISSUE #11102: add socialnetwork as referrertype in all referrers report
In dashboard and referrers menu, there is a report based on social networks. When looking at the Referrer Types report, we miss these items.
By adding the social networks as a type, with there sub-items, we get a more complete overview of referrer types.

### PR DESCRIPTION
**Pull request by sgiehl**
fixes #11102

### CODE REVIEW

**Comment by tsteur** [comment]
FYI: We'll need to check and adjust some plugins as well (not part of this PR) like Users Flow, Attribution, Funnels, ...

I presume people will see afterwards a drop in websites referrer types? I wonder if we need to mention in a blog post and explain they are now attributed differently?

**Comment by sgiehl** [comment]
> I presume people will see afterwards a drop in websites referrer types?

I'm wondering if it maybe would make sense to automatically add an annotation to referrer type report on the day of update to point that out?

**Comment by sgiehl** [comment]
Ready for a first review. Open todos:
- [ ] Update CHANGELOG
- [ ] Maybe add update script to add annotation
- [ ] Update UI files and maybe OmniFixture

**Comment by diosmosis** [comment]
For dates before this goes into effect, I see that 'Social Networks' isn't added to the "Referrer Types" report:

(old period)
image

(new period)
image

**Comment by diosmosis** [comment]
Should there be a reattribution command for users that want to apply this to old data (eg, in order to segment by social network in old data)? I suppose they could segment by referrer URL, but I'm not so sure. (cc @mattab)

**Comment by diosmosis** [comment]
Probably doesn't have to be fixed by this PR, but I notice in the "all referrers" report, social networks are links, but websites aren't:

image

Seems inconsistent.

**Comment by diosmosis** [comment]
In the visitor log the social referrer is a link, but in the visitor profile it's not:

image

Should it be a link there?

**Review by diosmosis — COMMENTED** [review]
Left some comments.

**Comment by mattab** [comment]
> Should there be a reattribution command for users that want to apply this to old data (eg, in order to segment by social network in old data)? I suppose they could segment by referrer URL, but I'm not so sure. 

maybe this could be a separate issue, but not needed for now to implement it.

**Comment by sgiehl** [comment]
I've updated the code according to the review comments.

> For dates before this goes into effect, I see that 'Social Networks' isn't added to the "Referrer Types" report:

That's not easily doable without reprocessing the reports, as the old reports for referrer type did not include that metric. Instead the social networks were included in the websites metric.

**Comment by diosmosis** [comment]
> That's not easily doable without reprocessing the reports, as the old reports for referrer type did not include that metric. Instead the social networks were included in the websites metric.

I thought you could do the same thing for when there is no social data in `Referrers.getSocials` in `Referrers.getReferrerType`. Ie, if `$idSubtable == REFERRER_TYPE_SOCIAL` and there is no data returned by the `getSocials` method, then build the datatable from the `getWebsites()` result. I guess you would also have to remove social URLs from the getWebsites() result if `$idSubtable == REFERRER_TYPE_WEBSITE`. I guess this isn't doable? Of course, I haven't thought it through to the point that I know it'll work.

**Comment by sgiehl** [comment]
> I thought you could do the same thing for when there is no social data in Referrers.getSocials in Referrers.getReferrerType. Ie, if $idSubtable == REFERRER_TYPE_SOCIAL and there is no data returned by the getSocials method, then build the datatable from the getWebsites() result. I guess you would also have to remove social URLs from the getWebsites() result if $idSubtable == REFERRER_TYPE_WEBSITE. I guess this isn't doable? Of course, I haven't thought it through to the point that I know it'll work.

Sorry for the delay. The referrer type numbers shown in that table are stored in an extra report. Splitting up the number for websites and socials for older data is not easy. We would need to query the reports for websites and split it up to websites and socials and then calculate all the numbers based on those resulting reports. I think that might be getting very slow when e.g. viewing row evolution. Also those referrer type reports are calculated as goal reports. I guess recalculating those would only be possible by reprocessing the logs.
Also the resulting numbers might differ if e.g. a website has more referrers than the report limit allows. While the calculated type report would show the correct number, the websites reports would only contain the limited dataset...

**Review by diosmosis — APPROVED** [review]
Ok, thought it might be too difficult to be worth it.

Will leave final review/merge to @mattab / @tsteur

File: plugins/Referrers/API.php [hunk 1]
```diff
@@ -328,22 +330,88 @@ public function getUrlsFromWebsiteId($idSite, $period, $date, $idSubtable, $segm
...
+        $dataTable = $this->completeSocialTablesWithOldReports($dataTable, $idSite, $period, $date, $segment, $expanded, $flat);
+
+        $dataTable->filter('MetadataCallbackAddMetadata', array('url', 'logo', function ($url) { return Social::getInstance()->getLogoFromUrl($url); }));
+
+        return $dataTable;
+    }
+
+    private function completeSocialTablesWithOldReports($dataTable, $idSite, $period, $date, $segment, $expanded, $flat)
+    {
+        return $this->combineDataTables($dataTable, function() use ($idSite, $period, $date, $segment, $expanded, $flat) {
```
Comment (tsteur) [inline plugins/Referrers/API.php]: Maybe we could just add a simple comment to state that we may want to remove this in a year or more likely two years as it then not really won't be needed anymore? With a link to this issue...

File: plugins/Referrers/API.php [hunk 2]
```diff
@@ -328,22 +330,88 @@ public function getUrlsFromWebsiteId($idSite, $period, $date, $idSubtable, $segm
...
+        return $dataTable;
+    }
 
-        $dataTable->filter('ColumnCallbackDeleteRow', array('label', function ($url) { return !Social::getInstance()->isSocialUrl($url); }));
-        $dataTable->filter('ColumnCallbackAddMetadata', array('label', 'url', function ($url) { return Social::getInstance()->getMainUrl($url); }));
-        $dataTable->filter('GroupBy', array('label', function ($url) { return Social::getInstance()->getSocialNetworkFromDomain($url); }));
+    /**
+     * @param DataTable $dataTable
+     */
+    protected function filterWebsitesForSocials($dataTable, $idSite, $period, $date, $segment, $expanded, $flat)
```
Comment (tsteur) [inline plugins/Referrers/API.php]: can we make those methods private or are there tests for it?

File: plugins/Referrers/API.php [hunk 3]
```diff
@@ -328,22 +330,88 @@ public function getUrlsFromWebsiteId($idSite, $period, $date, $idSubtable, $segm
...
+    {
+        $dataTable->filter('ColumnCallbackDeleteRow', array('label', function ($url) {
```
Comment (tsteur) [inline plugins/Referrers/API.php]: Might be good to check performance at some stage with a profiler on real data ideally just to see... especially re the groupby but also everything else like those filters.

File: plugins/Referrers/API.php (hunk 1 above)
Comment (tsteur) [inline plugins/Referrers/API.php]: Is that only needed for DataTableMaps? I see below in `combineDataTables ` we handle it also when it is a regular data table but not sure when this is needed?

Comment (tsteur) [reply plugins/Referrers/API.php]: I see it is when data is stored the old way...

File: plugins/Referrers/API.php [hunk 4]
```diff
@@ -328,22 +330,88 @@ public function getUrlsFromWebsiteId($idSite, $period, $date, $idSubtable, $segm
...
+
+    protected function combineDataTables($dataTable, $callbackForAdditionalData)
+    {
+        $isMap = false;
+        $hasEmptyTable = false;
+        if ($dataTable instanceof DataTable\Map) {
+            $isMap = true;
+            $dataTables = $dataTable->getDataTables();
+        } else {
+            $dataTables = [$dataTable];
```
Comment (tsteur) [inline plugins/Referrers/API.php]: I wonder if the code could be simplified there eg just `if (!$dataTable->getRowCount()) return $callbackForAdditionalData();`? Bit hard to understand currently with if's in top and bottom

**Comment by tsteur** [comment]
Left a few minor comments... we will then definitely need to check which plugins need updating. I think eg funnels, multi attribution ...

**Comment by sgiehl** [comment]
@tsteur I've improved the code according to your suggestions. I've not yet done a profiling and don't know when I will find some time to do that properly. 
But actually the filters aren't new. Those filters were used before to generate the social networks reports on the fly as well...

**Comment by tsteur** [comment]
Should be all fine then @sgiehl if they were all there before. If there are new ones, be good to have a look just to be sure.

**Pull request merged**